core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -154,6 +163,162 @@ class ControllerApis(val requestChannel: RequestChannel,
+            "Neither topic name nor id were specified."))
+        } else {
+          maybeAppendToIdsToResolve(topic.topicId())
+        }
+      } else {
+        if (topic.topicId().equals(ZERO_UUID)) {
+          maybeAppendToTopicNamesToResolve(topic.name())
+        } else {
+          appendResponse(topic.name(), topic.topicId(), new ApiError(INVALID_REQUEST,
+            "You may not specify both topic name and topic id."))
+        }
+      }
+    }
+
+    val idToName = new util.HashMap[Uuid, String]
+    val unknownTopicNameErrors = new util.HashMap[String, ApiError]
+    def maybeAppendToIdToName(id: Uuid, name: String): Unit = {
+      if (duplicatedTopicIds.contains(id) || idToName.put(id, name) != null) {
+          appendResponse(name, id, new ApiError(INVALID_REQUEST,
+              "The same topic was specified by name and by id."))
+          idToName.remove(id)
+          duplicatedTopicIds.add(id)
+      }
+    }
+    controller.findTopicIds(topicNamesToResolve).get().asScala.forKeyValue {
+      case (name, idOrError) => if (idOrError.isError) {
+        unknownTopicNameErrors.put(name, idOrError.error())
+      } else {
+        maybeAppendToIdToName(idOrError.result(), name)
+      }
+    }
+
+    /**
+     * There are 5 error cases to handle here:
+     *
+     * 1. name provided, topic exists, can't delete, maybe can describe => TOPIC_AUTHORIZATION_FAILED
+     * 2. name provided, topic doesn't exist, can't delete, can describe => UNKNOWN_TOPIC_OR_PARTITION
+     * 3. name provided, topic doesn't exist, can't delete, can't describe => TOPIC_AUTHORIZATION_FAILED
+     * 4. id provided, topic exists, can't delete, maybe can describe => TOPIC_AUTHORIZATION_FAILED
+     * 5. id provided, topic doesn't exists, can't delete, maybe can describe => UNKNOWN_TOPIC_ID
+     */
+    controller.findTopicNames(topicIdsToResolve).get().asScala.forKeyValue {
+      case (id, nameOrError) => if (nameOrError.isError) {
+        // Case 5: we can't resolve the given topic ID to a name to feed to the
+        // Authorizer.  Return UNKNOWN_TOPIC_ID.
+        appendResponse(null, id, nameOrError.error())
+      } else {
+        maybeAppendToIdToName(id, nameOrError.result())
+      }
+    }
+    if (!hasClusterAuth) {
+      val allTopicNames = idToName.values().asScala ++ unknownTopicNameErrors.keySet().asScala
+      val authorizedDescribeTopics = getDescribableTopics(allTopicNames)
+      val authorizedDeleteTopics = getDeletableTopics(allTopicNames)
+      val iterator = idToName.entrySet().iterator()
+      while (iterator.hasNext) {
+        val entry = iterator.next()
+        val topicName = entry.getValue
+        if (!authorizedDeleteTopics.contains(topicName)) {
+          // Case 1 or case 4: the topic exists, but we don't have permission to delete it.
+          val topicId = entry.getKey
+          if (topicIdsToResolve.contains(topicId)) {
+            appendResponse(null, topicId, new ApiError(TOPIC_AUTHORIZATION_FAILED))
```

Comment:
Could you add comment to explain why `name` must be null for this case?

Reply:
I think we need a case for describe vs. delete permissions right? Will a client know they have describe vs. delete permissions?